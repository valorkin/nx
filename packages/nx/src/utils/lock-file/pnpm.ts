import { LockFileData, PackageDependency } from './lock-file-type';
import { load, dump } from '@zkochan/js-yaml';
import { sortObject, hashString } from './utils';

type PackageMeta = {
  key: string;
  specifier?: string;
  isDevDependency?: boolean;
  isDependency?: boolean;
  dependencyDetails: Record<string, Record<string, string>>;
};

type Dependencies = Record<string, Omit<PackageDependency, 'packageMeta'>>;

type VersionInfoWithInlineSpecifier = {
  version: string;
  specifier: string;
};

type PnpmLockFile = {
  lockfileVersion: string;
  specifiers?: Record<string, string>;
  dependencies?: Record<
    string,
    string | { version: string; specifier: string }
  >;
  devDependencies?: Record<
    string,
    string | { version: string; specifier: string }
  >;
  packages: Dependencies;
};

const LOCKFILE_YAML_FORMAT = {
  blankLines: true,
  lineWidth: 1000,
  noCompatMode: true,
  noRefs: true,
  sortKeys: false,
};

/**
 * Parses pnpm-lock.yaml file to `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
export function parsePnpmLockFile(lockFile: string): LockFileData {
  const data: PnpmLockFile = load(lockFile);
  const { dependencies, devDependencies, packages, specifiers, ...metadata } =
    data;

  return {
    dependencies: mapPackages(
      dependencies,
      devDependencies,
      specifiers,
      packages,
      metadata.lockfileVersion.toString().endsWith('inlineSpecifiers')
    ),
    lockFileMetadata: { ...metadata },
    hash: hashString(lockFile),
  };
}

function mapPackages(
  dependencies: Record<string, string | VersionInfoWithInlineSpecifier>,
  devDependencies: Record<string, string | VersionInfoWithInlineSpecifier>,
  specifiers: Record<string, string>,
  packages: Dependencies,
  inlineSpecifiers: boolean
): LockFileData['dependencies'] {
  const mappedPackages: LockFileData['dependencies'] = {};

  Object.entries(packages).forEach(([key, value]) => {
    const { resolution, engines, ...dependencyDetails } = value;

    // construct packageMeta object
    const meta = mapMetaInformation(
      { dependencies, devDependencies, specifiers },
      inlineSpecifiers,
      key,
      dependencyDetails
    );

    // create new key
    const version = key.split('/').pop().split('_')[0];
    const packageName = key.slice(1, key.lastIndexOf('/'));
    const newKey = `${packageName}@${version}`;

    if (!mappedPackages[packageName]) {
      mappedPackages[packageName] = {};
    }
    if (mappedPackages[packageName][newKey]) {
      mappedPackages[packageName][newKey].packageMeta.push(meta);
    } else {
      mappedPackages[packageName][newKey] = {
        resolution,
        engines,
        version,
        packageMeta: [meta],
      };
    }
  });
  return mappedPackages;
}

// maps packageMeta based on dependencies, devDependencies and (inline) specifiers
function mapMetaInformation(
  {
    dependencies,
    devDependencies,
    specifiers,
  }: Omit<PnpmLockFile, 'lockfileVersion' | 'packages'>,
  hasInlineSpefiers,
  key: string,
  dependencyDetails: Record<string, Record<string, string>>
): PackageMeta {
  const matchingVersion = key.split('/').pop();
  const packageName = key.slice(1, key.lastIndexOf('/'));

  const isDependency = isVersionMatch(
    dependencies?.[packageName],
    matchingVersion,
    hasInlineSpefiers
  );
  const isDevDependency = isVersionMatch(
    devDependencies?.[packageName],
    matchingVersion,
    hasInlineSpefiers
  );

  let specifier;
  if (isDependency || isDevDependency) {
    if (hasInlineSpefiers) {
      specifier =
        getSpecifier(dependencies?.[packageName]) ||
        getSpecifier(devDependencies?.[packageName]);
    } else {
      if (isDependency || isDevDependency) {
        specifier = specifiers[packageName];
      }
    }
  }

  return {
    key,
    isDependency,
    isDevDependency,
    specifier,
    dependencyDetails,
  };
}

// version match for dependencies w/ or w/o inline specifier
function isVersionMatch(
  versionInfo: string | { version: string; specifier: string },
  matchingVersion,
  hasInlineSpefiers
): boolean {
  if (!versionInfo) {
    return false;
  }
  if (!hasInlineSpefiers) {
    return versionInfo === matchingVersion;
  }

  return (
    (versionInfo as VersionInfoWithInlineSpecifier).version === matchingVersion
  );
}

function getSpecifier(
  versionInfo: string | { version: string; specifier: string }
): string {
  return (
    versionInfo && (versionInfo as VersionInfoWithInlineSpecifier).specifier
  );
}

/**
 * Generates pnpm-lock.yml file from `LockFileData` object
 *
 * @param lockFile
 * @returns
 */
export function stringifyPnpmLockFile(lockFileData: LockFileData): string {
  const pnpmLockFile = unmapLockFile(lockFileData);

  return dump(pnpmLockFile, LOCKFILE_YAML_FORMAT);
}

// revert lock file to it's original state
function unmapLockFile(lockFileData: LockFileData): PnpmLockFile {
  const devDependencies: Record<
    string,
    string | VersionInfoWithInlineSpecifier
  > = {};
  const dependencies: Record<string, string | VersionInfoWithInlineSpecifier> =
    {};
  const packages: Dependencies = {};
  const specifiers: Record<string, string> = {};
  const inlineSpecifiers = lockFileData.lockFileMetadata.lockfileVersion
    .toString()
    .endsWith('inlineSpecifiers');

  const packageNames = Object.keys(lockFileData.dependencies);
  for (let i = 0; i < packageNames.length; i++) {
    const packageName = packageNames[i];
    const versions = Object.values(lockFileData.dependencies[packageName]);

    versions.forEach(({ packageMeta, resolution, engines }) => {
      (packageMeta as PackageMeta[]).forEach((meta) => {
        const { key, specifier } = meta;

        let version;
        if (inlineSpecifiers) {
          version = { specifier, version: key.slice(key.lastIndexOf('/') + 1) };
        } else {
          version = key.slice(key.lastIndexOf('/') + 1);

          if (specifier) {
            specifiers[packageName] = specifier;
          }
        }

        if (meta.isDependency) {
          dependencies[packageName] = version;
        }
        if (meta.isDevDependency) {
          devDependencies[packageName] = version;
        }
        packages[key] = {
          resolution,
          engines,
          ...meta.dependencyDetails,
        };
      });
    });
  }

  return {
    ...(lockFileData.lockFileMetadata as { lockfileVersion: string }),
    specifiers: sortObject(specifiers),
    dependencies: sortObject(dependencies),
    devDependencies: sortObject(devDependencies),
    packages: sortObject(packages),
  };
}

/**
 * Prunes the lock file data based on the list of packages and their transitive dependencies
 *
 * @param lockFileData
 * @returns
 */
export function prunePnpmLockFile(
  lockFileData: LockFileData,
  packages: string[]
): LockFileData {
  // todo(meeroslav): This functionality has not been implemented yet
  console.warn(
    'Pruning pnpm-lock.yaml is not yet implemented. Returning entire lock file'
  );
  return lockFileData;
}
