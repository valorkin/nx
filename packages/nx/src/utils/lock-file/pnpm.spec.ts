import { parsePnpmLockFile, stringifyPnpmLockFile } from './pnpm';
import {
  lockFile,
  lockFileWithInlineSpecifiers,
} from './__fixtures__/pnpm.lock';

describe('pnpm LockFile utility', () => {
  describe('standard lock file', () => {
    const parsedLockFile = parsePnpmLockFile(lockFile);

    it('should parse lockfile correctly', () => {
      expect(parsedLockFile.lockFileMetadata).toEqual({ lockfileVersion: 5.4 });
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(324);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
    });

    it('should map various versions of packages', () => {
      expect(
        Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping'])
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ]
      ).toBeDefined();
    });

    it('should map various instances of the same version', () => {
      const jestResolveDependency =
        parsedLockFile.dependencies['jest-pnp-resolver'][
          'jest-pnp-resolver@1.2.2'
        ];

      expect(jestResolveDependency.packageMeta.length).toEqual(2);
      expect((jestResolveDependency.packageMeta[0] as any).key).toEqual(
        '/jest-pnp-resolver/1.2.2_jest-resolve@28.1.1'
      );
      expect((jestResolveDependency.packageMeta[1] as any).key).toEqual(
        '/jest-pnp-resolver/1.2.2_jest-resolve@28.1.3'
      );

      expect(
        (jestResolveDependency.packageMeta[0] as any).dependencyDetails
          .dependencies
      ).toEqual({ 'jest-resolve': '28.1.1' });
      expect(
        (jestResolveDependency.packageMeta[1] as any).dependencyDetails
          .dependencies
      ).toEqual({ 'jest-resolve': '28.1.3' });
    });

    it('should properly extract specifier', () => {
      expect(
        (
          parsedLockFile.dependencies['@ampproject/remapping'][
            '@ampproject/remapping@2.2.0'
          ].packageMeta[0] as any
        ).specifier
      ).toBeUndefined();
      expect(
        (
          parsedLockFile.dependencies['typescript']['typescript@4.8.3']
            .packageMeta[0] as any
        ).specifier
      ).toEqual('~4.8.2');
    });

    it('should properly extract dev dependency', () => {
      expect(
        (
          parsedLockFile.dependencies['@ampproject/remapping'][
            '@ampproject/remapping@2.2.0'
          ].packageMeta[0] as any
        ).isDevDependency
      ).toEqual(false);
      expect(
        (
          parsedLockFile.dependencies['typescript']['typescript@4.8.3']
            .packageMeta[0] as any
        ).isDevDependency
      ).toEqual(true);
    });

    it('should match the original file on stringification', () => {
      expect(stringifyPnpmLockFile(parsedLockFile)).toEqual(lockFile);
    });
  });

  describe('lock file with inline specifiers', () => {
    const parsedLockFile = parsePnpmLockFile(lockFileWithInlineSpecifiers);

    it('should parse lockfile (IS)', () => {
      expect(parsedLockFile.lockFileMetadata).toEqual({
        lockfileVersion: '5.4-inlineSpecifiers',
      });
      expect(Object.keys(parsedLockFile.dependencies).length).toEqual(324);
      expect(
        parsedLockFile.dependencies['@ampproject/remapping']
      ).toMatchSnapshot();
      expect(parsedLockFile.dependencies['typescript']).toMatchSnapshot();
    });

    it('should map various versions of packages (IS)', () => {
      expect(
        Object.keys(parsedLockFile.dependencies['@jridgewell/gen-mapping'])
          .length
      ).toEqual(2);
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.1.1'
        ]
      ).toBeDefined();
      expect(
        parsedLockFile.dependencies['@jridgewell/gen-mapping'][
          '@jridgewell/gen-mapping@0.3.2'
        ]
      ).toBeDefined();
    });

    it('should map various instances of the same version (IS)', () => {
      const jestResolveDependency =
        parsedLockFile.dependencies['jest-pnp-resolver'][
          'jest-pnp-resolver@1.2.2'
        ];

      expect(jestResolveDependency.packageMeta.length).toEqual(2);
      expect((jestResolveDependency.packageMeta[0] as any).key).toEqual(
        '/jest-pnp-resolver/1.2.2_jest-resolve@28.1.1'
      );
      expect((jestResolveDependency.packageMeta[1] as any).key).toEqual(
        '/jest-pnp-resolver/1.2.2_jest-resolve@28.1.3'
      );

      expect(
        (jestResolveDependency.packageMeta[0] as any).dependencyDetails
          .dependencies
      ).toEqual({ 'jest-resolve': '28.1.1' });
      expect(
        (jestResolveDependency.packageMeta[1] as any).dependencyDetails
          .dependencies
      ).toEqual({ 'jest-resolve': '28.1.3' });
    });

    it('should properly extract specifier (IS)', () => {
      expect(
        (
          parsedLockFile.dependencies['@ampproject/remapping'][
            '@ampproject/remapping@2.2.0'
          ].packageMeta[0] as any
        ).specifier
      ).toBeUndefined();
      expect(
        (
          parsedLockFile.dependencies['typescript']['typescript@4.8.3']
            .packageMeta[0] as any
        ).specifier
      ).toEqual('~4.8.2');
    });

    it('should properly extract dev dependency (IS)', () => {
      expect(
        (
          parsedLockFile.dependencies['@ampproject/remapping'][
            '@ampproject/remapping@2.2.0'
          ].packageMeta[0] as any
        ).isDevDependency
      ).toEqual(false);
      expect(
        (
          parsedLockFile.dependencies['typescript']['typescript@4.8.3']
            .packageMeta[0] as any
        ).isDevDependency
      ).toEqual(true);
    });

    it('should match the original file on stringification (IS)', () => {
      expect(stringifyPnpmLockFile(parsedLockFile)).toEqual(
        lockFileWithInlineSpecifiers
      );
    });
  });
});
