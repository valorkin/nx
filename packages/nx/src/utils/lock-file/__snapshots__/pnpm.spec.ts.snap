// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnpm LockFile utility lock file with inline specifiers should parse lockfile (IS) 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "engines": Object {
      "node": ">=6.0.0",
    },
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {
          "dependencies": Object {
            "@jridgewell/gen-mapping": "0.1.1",
            "@jridgewell/trace-mapping": "0.3.15",
          },
          "dev": true,
        },
        "isDependency": false,
        "isDevDependency": false,
        "key": "/@ampproject/remapping/2.2.0",
        "specifier": undefined,
      },
    ],
    "resolution": Object {
      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    },
    "version": "2.2.0",
  },
}
`;

exports[`pnpm LockFile utility lock file with inline specifiers should parse lockfile (IS) 2`] = `
Object {
  "typescript@4.8.3": Object {
    "engines": Object {
      "node": ">=4.2.0",
    },
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {
          "dev": true,
          "hasBin": true,
        },
        "isDependency": false,
        "isDevDependency": true,
        "key": "/typescript/4.8.3",
        "specifier": "~4.8.2",
      },
    ],
    "resolution": Object {
      "integrity": "sha512-goMHfm00nWPa8UvR/CPSvykqf6dVV8x/dp0c5mFTMTIu0u0FlGWRioyy7Nn0PGAdHxpJZnuO/ut+PpQ8UiHAig==",
    },
    "version": "4.8.3",
  },
}
`;

exports[`pnpm LockFile utility standard lock file should parse lockfile correctly 1`] = `
Object {
  "@ampproject/remapping@2.2.0": Object {
    "engines": Object {
      "node": ">=6.0.0",
    },
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {
          "dependencies": Object {
            "@jridgewell/gen-mapping": "0.1.1",
            "@jridgewell/trace-mapping": "0.3.15",
          },
          "dev": true,
        },
        "isDependency": false,
        "isDevDependency": false,
        "key": "/@ampproject/remapping/2.2.0",
        "specifier": undefined,
      },
    ],
    "resolution": Object {
      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
    },
    "version": "2.2.0",
  },
}
`;

exports[`pnpm LockFile utility standard lock file should parse lockfile correctly 2`] = `
Object {
  "typescript@4.8.3": Object {
    "engines": Object {
      "node": ">=4.2.0",
    },
    "packageMeta": Array [
      Object {
        "dependencyDetails": Object {
          "dev": true,
          "hasBin": true,
        },
        "isDependency": false,
        "isDevDependency": true,
        "key": "/typescript/4.8.3",
        "specifier": "~4.8.2",
      },
    ],
    "resolution": Object {
      "integrity": "sha512-goMHfm00nWPa8UvR/CPSvykqf6dVV8x/dp0c5mFTMTIu0u0FlGWRioyy7Nn0PGAdHxpJZnuO/ut+PpQ8UiHAig==",
    },
    "version": "4.8.3",
  },
}
`;
